rfcs/NNNN-lit-3.0.md

```diff
@@ -0,0 +1,112 @@
+---
+Status: Active
+Champions: @justinfagnani
+PR: https://github.com/lit/rfcs/pull/16
+---
+
+# Lit 3.0
+
+A proposal for the next major release of Lit
+
+## Objective
+
+Allow us to make a few breaking changes that will let us reduce some engineering complexity, and very lightly reduce bundle size.
+
+### Goals
+- Require no code changes from developers using non-deprecated Lit 2.0 APIs
+- Drop IE support
+- Publish ES2021
+- Small cleanups (remove Lit pre-2.0 compatibility workarounds)
+
+### Non-Goals
+- Drop support for the Shady DOM polyfill
+- Not require additional polyfills for older browsers
+
+## Motivation
+
+### IE Support removal
+
+IE11 is now unsupported by Microsoft, most major websites, many frameworks, and apparently most of our users. The times we've asked our users if they need IE support, we haven't heard from anyone that they do at all. While this doesn't mean we don't have any users who require IE11 support, all signs point to it finally being time for us to move on.
+
+Doing so will let us turn down our flaky IE11 test runners and remove some IE-specific workarounds.
+
+### Publishing ES2021
+
+We published Lit 1.0 packages as ES2019 because of spotty browser support for some ES2021 features like logical assignment. The situation is much better now, and publishing native optional chaining and logical assignment expressions will give us a small bundle size and perf improvement. 
+
+### Clean up deprecated APIs
+
+We need to start the habit of regularly removing previously deprecated APIs to keep the codebase clean.
+
+## Detailed Design
+
+### IE11 removal
+
+- Turn down our IE test runners
+- Remove workaround for IE normalizing text nodes during clone ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/lit-html.ts#L974))
+- Simplify bound attribute handling ([source](https://github.com/lit/lit/blob/f2eb97962c7e77373b3b8861ab59639de22da3d0/packages/lit-html/src/lit-html.ts#L911))
```

Comment:
A few lines above the linked source here, there's a comment pointing out a `forEach` that's used instead of for-of for the purpose of avoiding compiler output for iteration. Is that something that we should consider rolling into this? (Or maybe queue up for a later major release, assuming we move to a faster / regular release schedule?)

Reply:
We should look into it, but maybe we just want to remove the comment... `.forEach()` just might be faster and more compact in some cases.

Reply:
I found more articles that indicate that `for..of` is faster, and a couple that claim `forEach()` is faster. I'll try to benchmark with tachometer, but all else being equal we would use `for..of` anyway, so I'll make the change (and thus require no or proper downleveling of for/of loops in 3.0). We can always back it out since then it's an implementation detail.